plugins/kubernetes-backend/src/service/KubernetesFetcher.ts

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  V1ConfigMap,
+  V1Deployment,
+  V1Pod,
+  V1ReplicaSet,
+  V1Secret,
+} from '@kubernetes/client-node';
+import { ClusterDetails } from '../cluster-locator/types';
+import { KubernetesClientProvider } from './KubernetesClientProvider';
+import { V1Service } from '@kubernetes/client-node/dist/gen/model/v1Service';
+import { Clients } from './types';
+import { Logger } from 'winston';
+
+export interface KubernetesFetcher {
+  fetchServicesByServiceId(
+    serviceId: string,
+    clusterDetails: ClusterDetails,
+  ): Promise<Array<V1Service>>;
+  fetchPodsByServiceId(
+    serviceId: string,
+    clusterDetails: ClusterDetails,
+  ): Promise<Array<V1Pod>>;
+  fetchConfigMapsByServiceId(
+    serviceId: string,
+    clusterDetails: ClusterDetails,
+  ): Promise<Array<V1ConfigMap>>;
+  fetchSecretsByServiceId(
+    serviceId: string,
+    clusterDetails: ClusterDetails,
+  ): Promise<Array<V1Secret>>;
+  fetchDeploymentsByServiceId(
+    serviceId: string,
+    clusterDetails: ClusterDetails,
+  ): Promise<Array<V1Deployment>>;
+  fetchReplicaSetsByServiceId(
+    serviceId: string,
+    clusterDetails: ClusterDetails,
+  ): Promise<Array<V1ReplicaSet>>;
+}
+
+export class KubernetesClientBasedFetcher implements KubernetesFetcher {
+  private readonly k8sClientProvider: KubernetesClientProvider;
+  private readonly logger: Logger;
+
+  constructor(k8sClientProvider: KubernetesClientProvider, logger: Logger) {
```

Comment:
Wondering if the could depend on a `KubernetesClusterLocator` instead of passing in the cluster details in every call?

Regarding the parameters, stick to options object instead of separate parameters unless it's very obvious stuff like `fromConfig`, and also then if you need to add more parameters, do it with an object, e.g. `fromConfig(config, {extraOption: true})`. With an options object you have much more information on the call site, especially for primitive literals, i.e. dafuq is `new Client(url, false, 10, true)`. But most importantly you have much better forwards compatibility if you end up needing to add more parameters.

Reply:
I actually wrote an implementation initially that used `KubernetesClusterLocator`. It turned out to then be very difficult to figure out what resources belonged to which cluster for the caller as all the pods from all clusters came back at the same time. 

I think it started to make a bit more sense when I changed to to this, as it removed a lot of logic which was just placing the correct resources in an object with the correct cluster.

Reply:
Ah alright, yeah this is nice and data driven so 👍 , just gonna need to keep an eye on it as we add auth options since that's in the cluster info atm

Reply:
> Regarding the parameters, stick to options object instead of separate parameters unless it's very obvious stuff like `fromConfig`, and also then if you need to add more parameters, do it with an object, e.g. `fromConfig(config, {extraOption: true})`. With an options object you have much more information on the call site, especially for primitive literals, i.e. dafuq is `new Client(url, false, 10, true)`. But most importantly you have much better forwards compatibility if you end up needing to add more parameters.

Repeating this bit :>

Reply:
Fixed in the latest commit